DirectProgramming/C++SYCL_FPGA/ReferenceDesigns/svd/src/svd.hpp

```diff
@@ -0,0 +1,262 @@
+  using RankDeficientFlagPipe =
+      sycl::ext::intel::pipe<IDRankDeficientFlagPipe, ac_int<1, false>, 3>;
+  using UTempMatrixPipe = sycl::ext::intel::pipe<IDUTempMatrixPipe, PipeType, 3>;
+  using UMatrixPipe = sycl::ext::intel::pipe<IDUMatrixPipe, PipeType, 3>;
+  using RMatrixPipe =
+      sycl::ext::intel::pipe<IDRMatrixPipe, T, kNumElementsPerDDRBurst * 4>;
+  using SMatrixPipe = sycl::ext::intel::pipe<IDSMatrixPipe, PipeType, 3>;
+  using VMatrixPipe = sycl::ext::intel::pipe<IDVMatrixPipe, PipeType, 3>;
+
+  // USM allocations
+  T *input_matrix_device;
+  ac_int<1, false> *rank_deficient_flag_device;
+  T *u_matrix_device;
+  T *s_matrix_device;
+  T *v_matrix_device;
+
+  if (q.get_device().has(sycl::aspect::usm_device_allocations)) {
+    std::cout << "Using device allocations" << std::endl;
+    // Allocate FPGA DDR memory.
+    input_matrix_device = sycl::malloc_device<T>(kInputMatrixSize, q);
+    rank_deficient_flag_device = sycl::malloc_device<ac_int<1, false>>(1, q);
+    u_matrix_device = sycl::malloc_device<T>(kUMatrixSize, q);
+    s_matrix_device = sycl::malloc_device<T>(kInputMatrixSize, q);
+    v_matrix_device = sycl::malloc_device<T>(kEigenVectorsMatrixSize, q);
+  } else if (q.get_device().has(sycl::aspect::usm_shared_allocations)) {
+    std::cout << "Using shared allocations" << std::endl;
+    // No device allocation support, using shared allocation instead 
+    input_matrix_device = sycl::malloc_shared<T>(kInputMatrixSize, q);
+    rank_deficient_flag_device = sycl::malloc_shared<ac_int<1, false>>(1, q);
+    u_matrix_device = sycl::malloc_shared<T>(kUMatrixSize, q);
+    s_matrix_device = sycl::malloc_shared<T>(kInputMatrixSize, q);
+    v_matrix_device = sycl::malloc_shared<T>(kEigenVectorsMatrixSize, q);
+  } else {
+    std::cerr << "USM device allocations or USM shared allocations must be "
+                 "supported to run this sample."
+              << std::endl;
+    std::terminate();
+  }
+
+  // Check that the malloc succeeded.
+  if (nullptr == input_matrix_device) {
+    std::cerr << "Error when allocating the input matrix." << std::endl;
+    std::terminate();
+  }
+  if (nullptr == rank_deficient_flag_device) {
+    std::cerr << "Error when allocating the rank deficient flag." << std::endl;
+    std::terminate();
+  }
+  if (nullptr == u_matrix_device) {
+    std::cerr << "Error when allocating the U matrix." << std::endl;
+    std::terminate();
+  }
+  if (nullptr == s_matrix_device) {
+    std::cerr << "Error when allocating the S matrix." << std::endl;
+    std::terminate();
+  }
+  if (nullptr == v_matrix_device) {
+    std::cerr << "Error when allocating the V matrix." << std::endl;
+    std::terminate();
+  }
+
+  // copy input matrix to memory that the device can access
+  q.memcpy(input_matrix_device, a_matrix.data(),
+           kInputMatrixSize * sizeof(float))
+      .wait();
+
+  // read input matrix from DDR
```

Comment:
Instead of having the kernel invocation read the same matrix over and over again, can we re-invoke the kernel multiple times with the same matrix? This invocation pattern is not realistic to what someone would do in real life. I don't think the _device code_ should be aware that it's always decomposing the same matrix over and over again, even if the _host code_ knows that it's getting the kernel to process the same input data over and over again.

I understand that you only have pre-computed singular values for this specific matrix, but you can either re-copy it multiple times, or place multiple copies of it in global memory and have the kernel grab the multiple matrices.

Reply:
This is what other samples do (this does not mean that this is the proper way).
I'd say that 1 matrix is suspiciously not enough and running 8 matrices in a repeat should be the sweet spot.

The device code is split into multiple parts, the only part that are aware of the repeat are the DMA kernels, which are not what the sample is demonstrating.
The assumption is that the user will use the headers that Kevin wrote to use in their own pipeline.
The DMA kernels being aware of the repeat is therefore a non-issue IMO.

Reply:
If the DMA kernels are aware of the repetitions, they are useless for real-world code, and we should clearly spell that out in the readme and in the comments on the DMA kernels. If I were looking at this reference design as a user, I would expect to be able to use it in my own application with my own data, with minimal changes to the host code. As it stands, we are expecting our users to make pretty major changes to the DMA kernels before this code is usable in a production design, and should be upfront about that.

If these types of 'repeating' DMA kernels are the only way to get good performance, then I would argue that these throughput numbers we are publishing are not representative of what a real user would expect in a real application.